### PR TITLE
Highlighter

### PR DESCRIPTION
**Pull request by mnunberg**

### CODE REVIEW

File: src/fragmenter.c [hunk 1]
```diff
@@ -0,0 +1,412 @@
+#include "fragmenter.h"
+#include "toksep.h"
+#include "tokenize.h"
+#include <ctype.h>
+#include <float.h>
+#include <sys/uio.h>
+#include <assert.h>
+
+#define myMin(a, b) (a) < (b) ? (a) : (b)
```
Comment (mnunberg) [inline src/fragmenter.c]: should really have these as universal macros somewhere, e.g. RSMin and RSMax?

File: src/fragmenter.c [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+  uint32_t termId;
+  int isStem = 0;
+
+  // See if this token matches any of our terms.
+  for (termId = 0; termId < fragList->numTerms; ++termId) {
+    const FragmentTerm *cur = fragList->terms + termId;
+    if (tokInfo->tokLen == cur->len && strncmp(tokInfo->tok, cur->tok, cur->len) == 0) {
+    } else if (tokInfo->stem && tokInfo->stemLen == cur->len &&
+               strncmp(tokInfo->stem, cur->tok, cur->len) == 0) {
+      isStem = 1;
```
Comment (mnunberg) [inline src/fragmenter.c]: TODO: isStem is not actually used anywhere for calculations and such. Should we care if it's a stem or not?

File: src/fragmenter.c [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    if (locs[jj].termId == termId) {
+      firstOcurrence = 0;
+      break;
+    }
+  }
+
+  if (firstOcurrence) {
+    curFrag->score += term->score;
+    // } else {
+    //   curFrag->score += term->score / 5;
```
Comment (mnunberg) [inline src/fragmenter.c]: what do you think? I think I commented it out because it was behaving weirdly.

File: src/fragmenter.c [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+  const char *preamble_s = NULL;
+
+  if (!preamble) {
+    preamble = &preamble_s;
+  }
+  if (!*preamble) {
+    *preamble = curFrag->buf;
+  }
+
+  for (size_t jj = 0; jj < nlocs; ++jj) {
```
Comment (mnunberg) [inline src/fragmenter.c]: ii

File: src/fragmenter.c [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    const TermLoc *curLoc = locs + jj;
+
+    size_t preambleLen = (curFrag->buf + curLoc->offset) - *preamble;
+
+    // Add any prior text
+    if (preambleLen) {
```
Comment (mnunberg) [inline src/fragmenter.c]: no need to check, since addToIov does the checks for us already.

File: src/fragmenter.c [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+  if (limitBefore == NULL) {
+    limitBefore = fragList->doc;
+  }
+  if (limitAfter == NULL) {
+    limitAfter = fragList->doc + fragList->docLen - 1;
+  }
+
+  // Subtract the number of context (i.e. non-match) words
+  // already inside the
```
Comment (mnunberg) [inline src/fragmenter.c]: formatting

File: src/fragmenter.c [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+  limitBefore = myMax(frag->buf - contextSize, limitBefore);
+  limitAfter = myMin(frag->buf + frag->len + contextSize, limitAfter);
+
+  before->iov_base = (void *)frag->buf;
+  before->iov_len = 0;
+
+  // Find the context immediately prior to our fragment, this means to advance
+  // the cursor as much as possible until a separator is reached, and then
+  // seek past that separator (if there are separators)
+  for (; limitBefore < frag->buf && !istoksep(*limitBefore); limitBefore++) {
```
Comment (mnunberg) [inline src/fragmenter.c]: i wonder how this works with Chinese

File: src/fragmenter.h [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+void FragmentList_HighlightWholeDocV(const FragmentList *fragList, const HighlightTags *tags,
+                                     Buffer *iovs);
+
+/** Highlight matches the entire document, returning it as a freeable NUL-terminated buffer */
+char *FragmentList_HighlightWholeDocS(const FragmentList *fragList, const HighlightTags *tags);
+
+/**
+ * Return fragments by their score. The highest ranked fragment is returned fist
+ */
+#define HIGHLIGHT_ORDER_SCORE 0x01
```
Comment (mnunberg) [inline src/fragmenter.h]: Should we make this into a proper enum?

File: src/query.c [hunk 9]
```diff
@@ -732,6 +755,9 @@ void Query_OnReopen(RedisModuleKey *k, void *privdata) {
   q->docTable = &sp->docs;
 }
 
+static void Query_NoteTerm(const RSIndexResult *res) {
```
Comment (mnunberg) [inline src/query.c]: Remove

File: src/query.c [hunk 10]
```diff
@@ -920,7 +946,179 @@ void QueryResult_Free(QueryResult *q) {
   free(q);
 }
 
-int QueryResult_Serialize(QueryResult *r, RedisSearchCtx *sctx, RSSearchRequest *req) {
+// Strip spaces from a buffer in place.
```
Comment (mnunberg) [inline src/query.c]: Might want to move this to RMUtil, or elsewhere?

File: src/query.c [hunk 11]
```diff
@@ -920,7 +946,179 @@ void QueryResult_Free(QueryResult *q) {
...
+      order = -1;
+      break;
+  }
+
+  iovsBufs = calloc(numIovBufs, sizeof(*iovsBufs));
+
+  FragmentList_HighlightFragments(&frags, &tags, fieldInfo->contextLen, iovsBufs, numIovBufs,
+                                  order);
+  if (fieldInfo->mode != SummarizeMode_Synopsis) {
+    printf("Replying with ARRAY..\n");
```
Comment (mnunberg) [inline src/query.c]: Remove

File: src/search_request.c [hunk 12]
```diff
@@ -48,6 +49,34 @@ RSSearchRequest *ParseRequest(RedisSearchCtx *ctx, RedisModuleString **argv, int
     req->slop = __INT_MAX__;
   }
 
+  int sumIdx;
```
Comment (mnunberg) [inline src/search_request.c]: I've redone all stateful parsing in another PR, but that's for after this is merged.

File: src/search_request.c [hunk 13]
```diff
@@ -179,6 +206,47 @@ RSSearchRequest *ParseRequest(RedisSearchCtx *ctx, RedisModuleString **argv, int
...
+  }
+
+  fields->fields = realloc(fields->fields, sizeof(*fields->fields) * ++fields->numFields);
+  ReturnedField *ret = fields->fields + (fields->numFields - 1);
+  memset(ret, 0, sizeof *ret);
+  ret->nameIndex = foundIndex;
+  return ret;
+}
+
+ReturnedField *FieldList_AddFieldR(FieldList *fields, RedisModuleString *s) {
```
Comment (mnunberg) [inline src/search_request.c]: Should we just use RedisModuleString because nothing is using const char directly?

File: src/search_request.h [hunk 14]
```diff
@@ -22,6 +22,51 @@ typedef enum {
...
+} ReturnedField;
+
+typedef struct {
+  char *openTag;
+  char *closeTag;
+
+  ReturnedField *fields;
+  size_t numFields;
+
+  char **rawFields;
```
Comment (mnunberg) [inline src/search_request.h]: Document why there are raw fields

File: src/tests/test_summarize.c [hunk 15]
```diff
@@ -0,0 +1,118 @@
+#include "test_util.h"
+#include "fragmenter.h"
+#include <stdio.h>
+#include "../rmutil/alloc.h"
+
+#define LOREM_IPSUM_FILE "./lorem_ipsum.txt"
```
Comment (mnunberg) [inline src/tests/test_summarize.c]: Remove. Not using lorem ipsum

File: src/fragmenter.c (hunk 1 above)
Comment (dvirsky) [reply src/fragmenter.c]: I'm generally using the one in sys/param.h

Comment (dvirsky) [reply src/fragmenter.c]: You can just create a minmax.h file with both macros. in util probably.

File: docs/Highlight.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+syntax, and the other is a more complicated one.
+
+### Simple Syntax
+
+```
+FT.SEARCH ... SUMMARIZE
+    [TAGS {open} {close}]
+    [FRAGSIZE {size}]
+    [NOTRUNCATE]
+    {num} {field}
```
Comment (dvirsky) [inline docs/Highlight.md]: Is the order important?

File: docs/Highlight.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+```
+
+* **TAGS** If passed, should specify the open and close tags to surround each
+    term with.
+    If not specified, then no special text will wrap the terms.
```
Comment (dvirsky) [inline docs/Highlight.md]: what is the default?

File: src/query.c [hunk 18]
```diff
@@ -195,6 +196,24 @@ void Query_SetIdFilter(Query *q, IdFilter *f) {
   Query_SetFilterNode(q, NewIdFilterNode(f));
 }
 
+static int maybeAddSumTerm(Query *q, const InvertedIndex *idx, const RSToken *tn, int needsCopy) {
```
Comment (dvirsky) [inline src/query.c]: Leaky abstraction alert :) I don't want the query to be aware of the fragmenter

File: src/query.c [hunk 19]
```diff
@@ -920,7 +946,179 @@ void QueryResult_Free(QueryResult *q) {
...
+  FragmentList frags;
+  const FragmentTerm *terms = (const FragmentTerm *)Buffer_GetData(&q->termsMeta);
+  size_t numTerms = BUFFER_GETSIZE_AS(&q->termsMeta, FragmentTerm);
+
+  // Dump the terms:
+  // for (size_t ii = 0; ii < numTerms; ++ii) {
+  //   printf("Have term '%.*s'\n", (int)terms[ii].len, terms[ii].tok);
+  // }
+
+  Stemmer *stemmer = NewStemmer(SnowballStemmer, q->language);
```
Comment (dvirsky) [inline src/query.c]: Can we avoid stemming in the highlighter? It's really slow. Doesn't the new offset based impl. allow removing this?

**Comment by mnunberg** [comment]
So this is the big bomb. This PR is where it all happens. It's still +600 lines, but that should be more manageable than several thousand :)

And now I'm off! See you!

File: src/extension.c [hunk 20]
```diff
@@ -142,7 +142,8 @@ void Ext_ExpandToken(struct RSQueryExpanderCtx *ctx, const char *str, size_t len
 
 /* Set the query payload */
 void Ext_SetPayload(struct RSQueryExpanderCtx *ctx, RSPayload payload) {
-  ctx->query->payload = payload;
+  // ctx->query->payload = payload;
+  printf("NO PAYLOAD SET!\n");
```
Comment (dvirsky) [inline src/extension.c]: My Bad - I forgot to reimplement it after refactor

File: src/highlight_processor.c [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+
+/**
+ * Attempts to fragmentize a single field from its offset entries. This takes
+ * the field name, gets the matching field ID, retrieves the offset iterator
+ * for the field ID, and fragments the text based on the offsets. The fragmenter
+ * itself is in fragmenter.{c,h}
+ *
+ * Returns true if the fragmentation succeeded, false otherwise.
+ */
+static int fragmentizeOffsets(const RSSearchRequest *req, const char *fieldName,
```
Comment (dvirsky) [inline src/highlight_processor.c]: Do not pass the whole request if you only need the spec from it. Also, the spec object might change on concurrent switches, you need to get the query execution context or the processor context here

File: src/highlight_processor.c [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+      }
+    } else {
+      isLastSpace = 0;
+      out[oix++] = s[ii];
+    }
+  }
+  return oix;
+}
+
+static void summarizeField(const RSSearchRequest *req, const ReturnedField *fieldInfo,
```
Comment (dvirsky) [inline src/highlight_processor.c]: again, do not pass the request here. you only need the spec to pass to fragmentizeOffsets

File: src/highlight_processor.c [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+  FragmentList_Free(&frags);
+}
+
+static int hlp_Next(ResultProcessorCtx *ctx, SearchResult *r) {
+  int rc = ResultProcessor_Next(ctx->upstream, r, 0);
+  if (rc == RS_RESULT_EOF) {
+    return rc;
+  }
+
+  RSSearchRequest *req = ctx->privdata;
```
Comment (dvirsky) [inline src/highlight_processor.c]: again, don't save the whole request, save the part you actually need

File: src/highlight_processor.c [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+static void hlp_Free(ResultProcessor *rp) {
+  free(rp);
+}
+
+ResultProcessor *NewHighlightProcessor(ResultProcessor *parent, RSSearchRequest *req) {
+  ResultProcessor *rp = NewResultProcessor(parent, req);
+  rp->Next = hlp_Next;
+  rp->Free = hlp_Free;
```
Comment (dvirsky) [inline src/highlight_processor.c]: RTFM - if Free is NULL we call free(rp) automatically, so no need to implement it if you don't have special logic

File: src/query_serializer.c [hunk 25]
```diff
@@ -0,0 +1,279 @@
+#if 0
```
Comment (dvirsky) [inline src/query_serializer.c]: remove this file if it's not needed anymore

**Pull request merged**